File: src/java/magma/Stream.java
package magma;

import java.util.function.Function;
import java.util.function.Predicate;

public interface Stream<Value> {
    static <T> Stream<T> of(final T... values) {
        return Stream.fromArray(values);
    }

    static <T> Stream<T> fromArray(final T[] values) {
        return new HeadedStream<>(new RangeHead(values.length)).map(index -> values[index]);
    }

    static <T> Stream<T> empty() {
        return new HeadedStream<>(new EmptyHead<>());
    }

    <R> Stream<R> map(Function<Value, R> mapper);

    <C> C collect(Collector<Value, C> collector);

    <R> Stream<R> flatMap(Function<Value, Stream<R>> mapper);

    Stream<Value> filter(Predicate<Value> predicate);

    Optional<Value> next();
}

File: src/node/magma/Stream.ts


interface Stream<Value> {
	fromArray<Some[value=]>() : Stream<Some[value=]> {
	}
	empty<Some[value=]>() : Stream<Some[value=]> {
	}
	map() : Stream<Some[value=]>;
	collect() : C;
	flatMap() : Stream<Some[value=]>;
	filter() : Stream<Some[value=]>;
	next() : Optional<Some[value=]>;
}

